sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,602 @@
+  }
+
+  uint32_t compressed_limit = 0;
+  compressed_limit |= (rx_level & SPI_DEVICE_FIFO_LEVEL_RXLVL_MASK)
+                      << SPI_DEVICE_FIFO_LEVEL_RXLVL_OFFSET;
+  compressed_limit |= (tx_level & SPI_DEVICE_FIFO_LEVEL_TXLVL_MASK)
+                      << SPI_DEVICE_FIFO_LEVEL_TXLVL_OFFSET;
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_FIFO_LEVEL_REG_OFFSET,
+                      compressed_limit);
+
+  return kDifSpiDeviceResultOk;
+}
+
+/**
+ * Parameters for compressing and decompressing a FIFO pointer register.
+ */
+typedef struct fifo_ptr_params {
+  ptrdiff_t offset;
+  ptrdiff_t write_offset;
+  ptrdiff_t read_offset;
+  uint32_t write_mask;
+  uint32_t read_mask;
+} fifo_ptr_params_t;
+
+static const fifo_ptr_params_t kTxFifoParams = {
+    .offset = SPI_DEVICE_TXF_PTR_REG_OFFSET,
+    .write_offset = SPI_DEVICE_TXF_PTR_WPTR_OFFSET,
+    .write_mask = SPI_DEVICE_TXF_PTR_WPTR_MASK,
+    .read_offset = SPI_DEVICE_TXF_PTR_RPTR_OFFSET,
+    .read_mask = SPI_DEVICE_TXF_PTR_RPTR_MASK,
+};
+
+static const fifo_ptr_params_t kRxFifoParams = {
+    .offset = SPI_DEVICE_RXF_PTR_REG_OFFSET,
+    .write_offset = SPI_DEVICE_RXF_PTR_WPTR_OFFSET,
+    .write_mask = SPI_DEVICE_RXF_PTR_WPTR_MASK,
+    .read_offset = SPI_DEVICE_RXF_PTR_RPTR_OFFSET,
+    .read_mask = SPI_DEVICE_RXF_PTR_RPTR_MASK,
+};
+
+/**
+ * An exploded FIFO pointer value, consisting of a |uint11_t|
+ * offset part, and a |uint1_t| phase.
+ */
+typedef struct fifo_ptr {
+  uint16_t offset;
+  bool phase;
+} fifo_ptr_t;
+
+// Masks for extracting the phase bit, and the address part, out of a FIFO
+// offset.
+static const uint16_t kPhaseMask = SPI_DEVICE_BUFFER_SIZE_BYTES;
+static const uint16_t kOffsetMask = SPI_DEVICE_BUFFER_SIZE_BYTES - 1;
+
+static void fifo_ptr_increment(fifo_ptr_t *ptr, uint16_t increment,
+                               uint16_t fifo_len) {
+  uint32_t inc_with_overflow = ptr->offset + increment;
+  // If we would overflow, wrap and flip the overflow bit.
+  if (inc_with_overflow >= fifo_len) {
+    inc_with_overflow -= fifo_len;
+    ptr->phase = !ptr->phase;
+  }
+
+  ptr->offset = inc_with_overflow & kOffsetMask;
+}
+
+/**
+ * A decompressed FIFO pointer register, consisting of a read offset
+ * and a write offset within the FIFO region.
+ *
+ * The offsets themselves are only |uint11_t|, with an additional
+ * 12th "phase" bit used for detecting the wrapparound behavior of
+ * the ring buffer FIFOs.
+ */
+typedef struct fifo_ptrs {
+  fifo_ptr_t write_ptr;
+  fifo_ptr_t read_ptr;
+} fifo_ptrs_t;
+
+static fifo_ptrs_t decompress_ptrs(const dif_spi_device_t *spi,
+                                   fifo_ptr_params_t params) {
+  uint32_t ptr = mmio_region_read32(spi->base_addr, params.offset);
+  uint16_t write_val =
+      (uint16_t)((ptr >> params.write_offset) & params.write_mask);
+  uint16_t read_val =
+      (uint16_t)((ptr >> params.read_offset) & params.read_mask);
+  return (fifo_ptrs_t){
+      .write_ptr =
```

Comment:
`.write_ptr =` -> `.write_ptr = {` ?

Reply:
Again, this is not what I wrote, but rather the formatter's doing.